### PR TITLE
[supervisor] avoid auto exposing the same port multiple times

### PR DESCRIPTION
**Pull request by akosyakov**
#### What it does

It can happen that exposed ports update takes long and meantime other configured/served updates can happen. It creates the pressure and can lead to poor ports availability. This PR makes sure that the same port can be auto exposed by configured/served ports only once if global port did not change, i.e. if local port was not bound/unbound to/from localhost.

#### How to test

- Tests should succeed.
- Start a workspace with many ports, try to toggle their visibility, change configs, serve/unserve the same port on localhost or globally. See how to test in https://github.com/gitpod-io/gitpod/pull/1909 for instance.

### CODE REVIEW

**Comment by svenefftinge** [comment]
Related? https://github.com/gitpod-io/gitpod/issues/2423

File: components/supervisor/pkg/ports/ports.go [hunk 1]
```diff
@@ -321,16 +328,21 @@ func (pm *Manager) nextState() map[uint32]*managedPort {
...
 }
 
+func (pm *Manager) autoExpose(ctx context.Context, mp *managedPort, public bool) {
+	err := pm.E.Expose(ctx, mp.LocalhostPort, mp.GlobalPort, public)
+	if err != nil {
+		log.WithError(err).WithField("port", *mp).Warn("cannot auto-expose port")
+		return
+	}
+	pm.autoExposed[mp.LocalhostPort] = mp.GlobalPort
+	log.WithField("port", *mp).Warn("auto-expose port")
```
Comment (svenefftinge) [inline components/supervisor/pkg/ports/ports.go:343]: should be Debug level

Comment (akosyakov) [reply components/supervisor/pkg/ports/ports.go:343]: would be Info fine? I wanted to see whether we are spamming or not

**Review by csweichel — APPROVED** [review]
Nice one. Changes LGTM

**Pull request merged**